matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/RoomService.kt

```diff
@@ -229,4 +229,12 @@ interface RoomService {
      */
     fun getFlattenRoomSummaryChildrenOfLive(spaceId: String?,
                                             memberships: List<Membership> = Membership.activeMemberships()): LiveData<List<RoomSummary>>
+
+    /**
+     * Refreshes the RoomSummary LatestPreviewContent for the given @param roomId
+     * If the roomId is null, all rooms are updated
+     *
+     * This is useful for refreshing summary content with encrypted messages after receiving new room keys
+     */
+    fun refreshJoinedRoomSummaryPreviews(roomId: String?)
```

Comment:
According to your video and following the implementation, the changes LGTM.
But I am wondering why we have to expose such API to the app. Can't this new decryption attempt be managed only SDK side? Since the SDK maintain the latestPreviewableEvent in the RoomSummary to me it would make more sense.

Reply:
moving to the sdk also makes sense to me, the tricky part is where~

in the timeline it's managed by the SDK `DefaultTimeline` abstraction which acts as the datasource and holder for the listeners to allow the SDK to manage the crypto/pagination callbacks, from the client `ViewModel` we simply call `timeline.start()`
 
currently the room summaries is just a data model which clients observe using `RoomService.getRoomSummariesLive`, we could add a `start/stop` to the `RoomService` to internally register for crypto callbacks or create another abstraction on top of the summaries similar to `Timeline` :thinking:  what do you think?

Reply:
using the `FlowSession` abstraction we can wrap and combine the room and crypto services 

```kotlin
fun liveRoomSummaries(queryParams: RoomSummaryQueryParams, sortOrder: RoomSortOrder = RoomSortOrder.NONE): Flow<List<RoomSummary>> {
    val refreshRoomSummariesOnCryptoSessionChange = object : NewSessionListener {
        override fun onNewSession(roomId: String?, senderKey: String, sessionId: String) {
            session.refreshJoinedRoomSummaryPreviews(roomId)
        }
    }

    return session.getRoomSummariesLive(queryParams, sortOrder).asFlow()
            .startWith(session.coroutineDispatchers.io) { session.getRoomSummaries(queryParams, sortOrder) }
            .onStart { session.cryptoService().addNewSessionListener(refreshRoomSummariesOnCryptoSessionChange) }
            .onCompletion { session.cryptoService().removeSessionListener(refreshRoomSummariesOnCryptoSessionChange) }
}
```

with the current keys being updated in background workers it makes it very difficult (and potentially easier to introduce memory leaks) to attach a key listener in the lower levels of the sync handler or dynamically spawning them when calling the room summary update

conceptually I like the idea of an abstraction on top of the Session api (similar to the `SessionFlow`) which would interact and coordinate with the separate services to provide a source of truth